### DISCUSSION TITLE
Porting LÖVE to PSP (again) and what is missing

### DISCUSSION THREAD

**Discussion by tpimh** [pspdev/pspdev#65] [Ideas]
LÖVE aka Love2D is a nice little game engine allowing you to implement your game almost entirely in Lua. It's easy to learn, fun to use, and the documentation is pretty okay. I have created a couple of small games with it, and I enjoyed the process. I believe, there were several attempts to port it to PSP, but all these ports are now abandoned. I would like to start a new port (possibly targeting not only the PSP, but also Vita, PS2, PS3, etc.) similar to [LovePotion](https://github.com/lovebrew/lovepotion) for Nintendo consoles.

**What are we missing?**

Official Love2D dependencies are:

- [x] [SDL2](https://pspdev.github.io/psp-packages/sdl2.html)
- [ ] OpenGL 2.1+ / OpenGL ES 2+ (we have [pspgl](https://pspdev.github.io/psp-packages/pspgl.html), but it's only targeting OpenGL 1.5 and OpenGL ES 1.1)
- [ ] OpenAL (we have [openal-soft](https://pspdev.github.io/psp-packages/openal.html), also [an unfinished port of MojoAL](https://github.com/stdgregwar/psp-packages/tree/mojoal/mojoal), and [ALdc port](https://gitlab.com/simulant/aldc) is looking promising)
- [x] Lua / LuaJIT / LLVM-lua (we have [Lua 5.1](https://pspdev.github.io/psp-packages/lua51.html), [Lua 5.2](https://pspdev.github.io/psp-packages/lua52.html), [Lua 5.3](https://pspdev.github.io/psp-packages/lua53.html), [Lua 5.4](https://pspdev.github.io/psp-packages/lua54.html), and [an unfinished port of LuaJIT](https://github.com/libcg/LuaJIT))
- [x] [FreeType](https://pspdev.github.io/psp-packages/freetype2.html) (and also [Harfbuzz](https://pspdev.github.io/psp-packages/harfbuzz.html))
- [x] [ModPlug](https://pspdev.github.io/psp-packages/libmodplug.html)
- [x] [Vorbisfile](https://pspdev.github.io/psp-packages/libvorbis.html)
- [x] [Theora](https://pspdev.github.io/psp-packages/theora.html)

**What will not compile?**

Additionally to dependencies, Love2D also bundles a lot of libraries that may not compile that easily:

- [ ] Wuff
- [ ] box2d (we have [a port](https://github.com/PSP-Archive/Box2D), but it needs to be revisited and updated)
- [ ] ddsparse
- [ ] dr
- [x] enet (we have [a port](https://pspdev.github.io/psp-packages/enet.html))
- [ ] glad
- [ ] glslang
- [ ] lodepng
- [x] lua53 (we have [a port](https://pspdev.github.io/psp-packages/lua53.html), but I guess it's not needed if Lua is used as an external dependency)
- [ ] luahttps
- [x] luasocket (we have [a port](https://pspdev.github.io/psp-packages/luasocket.html))
- [x] lz4 (we have [a port](https://pspdev.github.io/psp-packages/lz4.html))
- [ ] noise1234
- [x] physfs (we have [a port](https://pspdev.github.io/psp-packages/physfs))
- [ ] spirv_cross
- [x] stb (we have [a port](https://pspdev.github.io/psp-packages/stb.html))
- [ ] tinyexr
- [ ] utf8
- [ ] vma
- [ ] volk
- [ ] vulkanheaders
- [x] xxHash (we have [a port](https://pspdev.github.io/psp-packages/xxhash.html))

It's still to be investigated how much of the above "just works" and how much requires some patching.

Anyway, let me know what you think!

**Comment by sharkwouter** [comment]
This is cool! I've tried my hand at this before, but didn't get so far because back then we far fewer libraries.I think most of those should build just fine now today, though.

I don't think all libraries you listed are actually required. Most likely you can build without vulkanheaders and opengl. I think the main challenge will be physfs for sure. We don't have an actual port for that. We had a physfs port for a bit, but we removed it because it wasn't working iirc. The MR for that can be found here: https://github.com/pspdev/psp-packages/pull/32

**Reply by tpimh** [reply]
Yeah, there's no Vulkan on PSP, but including the bundled headers won't hurt. Let's see if pspgl is sufficient for all the graphics. Love2D has some shader support, but I think it can be disabled in the initial port. Not sure what LovePotion developers do on their hardware and how does it stand against PSP.

**Reply by sharkwouter** [reply]
It's hard to say now if porting LovePotion would be easier. It has a lot of Nintendo specific code.

**Reply by tpimh** [reply]
I'd like to keep LoveStation (or whatever it will be called) separate from LovePotion, but I can check how this or that part was ported to Nintendo platforms. Thankfully, the code is under MIT.

**Comment by xyzshantaram** [comment]
I'm actually working on getting general-purpose TLS sockets working on the PSP with the tlse library. I have no idea if that'll be helpful but if it will I would love to pitch it in.

**Comment by sharkwouter** [comment]
@tpimh can you add checkboxes for the libraries? We've since added some I think. At least xxhash was added by @diamant3

**Reply by tpimh** [reply]
Done. Also I updated all links to psp-packages repository while i was on it.

**Reply by diamant3** [reply]
I'll try to add some more packages according to that list. Anyway, hopefully we could have a Love2D port for PSP, guys ❤️. This is a really nice idea.

**Reply by diamant3** [reply]
I take the @stdgregwar' port of physfs, here pspdev/psp-packages#179. we should now have physfs❤️

**Comment by sharkwouter** [comment]
I did a little bit of testing and volk and vulkan can be excluded, which we should do.

A big problem we'll run into is that we have a 32 bit cpu, which means that lua does not have double support, which causes a ton of build issues. I'm not sure if we can really address this well. Perhaps we'll should do something ugly like `#define double float` and all unsigned ints will have to be changed to uint32 most likely. This will be a pain.

**Reply by tpimh** [reply]
Hmm, I think I have already seen a similar hack [somewhere](https://github.com/pspdev/psp-packages/blob/06f6c25c89ff966b4760045b4101ba3bf61d43b4/lua51/psp.patch#L251)... I think there should be a better solution, love2d runs fine on other 32bit CPUs.

**Comment by tpimh** [comment]
As a first step, it should be best to port [nLove](https://crankgaming.blogspot.com/2011/02/nlove-070-final-love2d-for-dingoo.html) which has much lower requirements. The code is old, and it's only providing a subset of a pretty old LÖVE API, but it still should be better than any of the existing implementations. It doesn't appear to be anywhere on GitHub, so I pushed it [here](https://github.com/tpimh/nlove/).

What is notable in this fork is the fact that OpenGL and OpenAL dependencies are made optional, and the code using them was rewritten using pure SDL.

It has less dependencies:
- [x] [Lua 5.1](https://pspdev.github.io/psp-packages/lua51.html)
- [x] [SDL 1.2](https://pspdev.github.io/psp-packages/sdl.html) + [SDL_image](https://pspdev.github.io/psp-packages/sdl-image.html) + [SDL_gfx](https://pspdev.github.io/psp-packages/sdl-gfx.html) + [SDL_mixer](https://pspdev.github.io/psp-packages/sdl-mixer.html)
- [x] [PhysicsFS](https://pspdev.github.io/psp-packages/physfs)
- [x] [Freetype](https://pspdev.github.io/psp-packages/freetype2.html)
- [x] [Vorbisfile](https://pspdev.github.io/psp-packages/libvorbis.html)

Also it bundles some of the dependencies that can possibly be replaced with ports in place or patched to convert to external dependencies:
- [ ] LuaSocket (I can't see the network module being included, so maybe it's not used)
- [ ] Box2D (the source seems pretty close to the one that was ported to PSP, but it's very hard to put an exact date on it)

**Reply by BrownNPC** [reply]
Please try compiling the newest Box2d version from source, it compiles very easily. I was able to compile Chipmunk2d aswell for my game. All I had to do was delete the code for Arm Neon , and it worked. @tpimh

**Reply by tpimh** [reply]
The most important thing that the port has, and the upstream lacks is the FPU optimizations, but these can be added easily (maybe even using [libpspla](https://github.com/Jayanky/pspla)?)

**Comment by sharkwouter** [comment]
So from the bundled libraries I am finding that box2d, ddparse, glad, lodepng, l4z, noise1234 and wuff do build during the build process.

I don't know what to do with glslang.

**Reply by diamant3** [reply]
What issue you got in glslang build?

**Reply by tpimh** [reply]
I'd say just patch it out. Last time I tried to compile it, I got [this problem](https://github.com/KhronosGroup/glslang/issues/3529), so I gave up.

**Reply by diamant3** [reply]
Sad, that was really annoying.

This is all the types used by the pspsdk If I'm correct. Here https://github.com/pspdev/pspsdk/blob/master/src%2Fbase%2Fpsptypes.h, can we modify it? instead of patching it out, patching that are tedious and exhausting. 😢

**Reply by sharkwouter** [reply]
It sounds more like the code of glsl is inconsistent and expects 2 different types to be the same on all systems. Should not be too hard to fix.

**Reply by stdgregwar** [reply]
Wouldn't it be "easier" or "cleaner" to just drop glsllang as a functionality from Love since there is anyway 0 chance that the PSP ever runs a shader ?

There must be a fair amount of patching shaders out of Love for it to run on PSP...

**Reply by sharkwouter** [reply]
Yeah, that does make sense. We'll not be using glsl at all.

**Reply by tpimh** [reply]
Can't say which is "easier" or "cleaner", but definitely the result will be "lighter", which is a good thing in most cases.

**Comment by sharkwouter** [comment]
I managed to get it to build, but it crashes upon startup, which I think happens because it tries to use a version of OpenGL which we do not have. Here is the code: https://github.com/sharkwouter/love/tree/psp-port2

Here is the comparison of what I changed:
https://github.com/love2d/love/compare/11.x...sharkwouter:love:psp-port2

You do have to update your lua library using `psp-pacman -Syu` for it to build, but it does build.

We can never get the shader stuff to build, that is for sure, but glsl is not optional and the minimum requirement for OpenGL is 2.1, which we cannot support because we have no hardware shader support.
